src/webgpu/api/operation/vertex_state.spec.ts

```diff
@@ -0,0 +1,366 @@
+    return indexBuffer;
+  }
+
+  run(
+    indexBuffer: GPUBuffer,
+    indexCount: number,
+    indexFormat: GPUIndexFormat,
+    indexOffset: number = 0,
+    primitiveTopology: GPUPrimitiveTopology = 'triangle-list'
+  ): GPUBuffer {
+    let pipeline: GPURenderPipeline;
+    // The indexFormat must be set in render pipeline descriptor that specifys a strip primitive
+    // topology for primitive restart testing
+    if (primitiveTopology === 'line-strip' || primitiveTopology === 'triangle-strip') {
+      pipeline = this.MakeRenderPipeline(primitiveTopology, indexFormat);
+    } else {
+      pipeline = this.MakeRenderPipeline(primitiveTopology);
+    }
+
+    const colorAttachment = this.device.createTexture({
+      format: kTextureFormat,
+      size: { width: kWidth, height: kHeight, depth: 1 },
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.OUTPUT_ATTACHMENT,
+    });
+
+    const result = this.device.createBuffer({
+      size: byteLength,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+
+    const vertexArray = new Float32Array(
+      /* prettier-ignore */ [
+      // float4 position
+      -1.0, 1.0, 0.0, 1.0,
+      1.0, -1.0, 0.0, 1.0,
+      1.0, 1.0, 0.0, 1.0,
+      -1.0, -1.0, 0.0, 1.0]
+    );
+
+    let bufferSize = vertexArray.byteLength;
+    // Create a large enough buffer for uint16 primitive restart value used as a big index.
+    if (
+      indexFormat === 'uint16' &&
+      primitiveTopology !== 'line-strip' &&
+      primitiveTopology !== 'triangle-strip'
+    ) {
+      bufferSize = 0xffff * 16;
+    }
+
+    const vertexBuffer = this.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.VERTEX,
+      mappedAtCreation: true,
+    });
+    new Float32Array(vertexBuffer.getMappedRange()).set(vertexArray);
+    vertexBuffer.unmap();
+
+    const encoder = this.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: colorAttachment.createView(),
+          loadValue: { r: 0.0, g: 0.0, b: 0.0, a: 1.0 },
+          storeOp: 'store',
+        },
+      ],
+    });
+    pass.setPipeline(pipeline);
+    pass.setVertexBuffer(0, vertexBuffer);
+    pass.setIndexBuffer(indexBuffer, indexFormat, indexOffset);
+    pass.drawIndexed(indexCount, 1, 0, 0, 0);
+    pass.endPass();
+    encoder.copyTextureToBuffer(
+      { texture: colorAttachment, origin: { x: 0, y: 0, z: 0 } },
+      { buffer: result, bytesPerRow, rowsPerImage },
+      { width: kWidth, height: kHeight, depth: 1 }
+    );
+    this.device.defaultQueue.submit([encoder.finish()]);
+
+    return result;
+  }
+
+  CreateExpectedUint8Array(renderShape: RenderShape): Uint8Array {
+    let texelValueBytes;
+    const arrayBuffer = new Uint8Array(byteLength);
+    for (let row = 0; row < kHeight; row++) {
+      for (let col = 0; col < kWidth; col++) {
+        if (
+          renderShape === RenderShape.Points &&
```

Comment:
Instead of this complicated logic, could be define an array of boolean for each shape and format them in a way that visually looks like the shape?